### PR TITLE
Getting tensor mechanics ready for parallel

### PR DESCRIPTION
**Pull request by tophmatthews**
ref #2975

### CODE REVIEW

**Comment by tophmatthews** [comment]
FYI @permcody

File: modules/tensor_mechanics/test/tests/crystal_plasticity/tests [hunk 1]
```diff
@@ -36,6 +36,7 @@
     input = 'crysp_cutback.i'
     exodiff = 'crysp_cutback_out.e'
     allow_warnings = true
+    max_parallel = 1
```
Comment (tophmatthews) [inline modules/tensor_mechanics/test/tests/crystal_plasticity/tests]: All of the tests in `tensor_mechanics/test/tests/crystal_plasticity` seem to be flagged for deprecation, but `FiniteStrainCrystalPlasticity` has not been deprecated. This is the only test that fails, so I just restrict it to be non-parallel only. I'm not sure if @permcody or @dschwen plan on yanking these out soon or not?

**Comment by permcody** [comment]
@tophmatthews - If you plan to switch the pc_type to `lu`, you really need to add `superlu = true` to the test spec. I suppose we need to actually test a configuration of PETSc where `superlu` isn't installed s you can see the problem. In that case, `lu` will complain that it is a serial only preconditioner. PETSc will run `superlu` when the number of CPUs is greater than one.

**Comment by permcody** [comment]
@milljm - Do we have any good way of running a version of PETSc without SuperLU? In Lieu of that, we could potentially add an extra PETSc config check to make sure that SuperLU is available when running in parallel.... That actually sounds a lot easier. @fdkong - Can you confirm the validity of my statement above?

**Comment by milljm** [comment]
@permcody we do not... All our scripts add that configuration switch. Looks like we need to provide some TLC to our scripts/update_and_rebuild_petsc.sh script. eg, allow a way to disable/add a configuration much like we would do with
```
update_and_rebuild_libmesh.sh --disable-shared --enable-static
```
for example.

**Comment by fdkong** [comment]
When getting any module parallel things in, it is better to test against `petsc-alt `as well.

**Comment by tophmatthews** [comment]
> If you plan to switch the pc_type to `lu`, you really need to add `superlu = true` to the test spec.

I'm not sure `lu` is necessary, rather that `asm` was giving different answers parallel/serial, not because it's necessary. Maybe there is a better one to use? I'll play around.

**Comment by tophmatthews** [comment]
> When getting any module parallel things in, it is better to test against `petsc-alt `as well.

Also, I found some tests failed depending on processor numbers, ie sometimes passed in -p2, but failed in -p3 or -p4.

Also, the heavy switch brought out some other failures.

**Comment by dschwen** [comment]
I'm phase field I'm seeing that I have to tighten tolerances quite a bit to
make parallel tests pass with asm.

**Comment by tophmatthews** [comment]
> I'm phase field I'm seeing that I have to tighten tolerances quite a bit to make parallel tests pass with asm.

Tolerances work for most of the failures, but then it becomes a balancing act to get convergence in serial, I couldn't find an appropriate balance with just tolerances.

**Comment by tophmatthews** [comment]
@permcody can you add TM to Partial Modules Parallel Test (In progress)?

**Comment by permcody** [comment]
TM is added. I'm ok with adding lu as a preconditioner. I just want to make sure that we don't pick up regressions for PETSc builds that don't have SuperLU or Mumps installed. The only robust way for me to do this is to add a target that actually tests that configuration. We'll be working on that!

**Comment by tophmatthews** [comment]
Removed lu where possible, and added `superlu = true` to the test spec when unavoidable.

**Comment by dschwen** [comment]
>  but then it becomes a balancing act to get convergence in serial

Are you regolding with tighter tolerances?

**Comment by tophmatthews** [comment]
> Are you regolding with tighter tolerances?

I was trying to avoid regolding? I didn't have to for some with tighter tolerances, and lu worked for the others

**Comment by tophmatthews** [comment]
Also, from looking at the exodiff, regolding wouldn't have worked, at least for one or two cases I was looking at.

**Comment by tophmatthews** [comment]
UGGGG SQA! Well if that's not motivation to not touch the `tests` file, I'm not sure what is

**Review by permcody — APPROVED** [review]

**Pull request merged**